### PR TITLE
VQVAE Implementation for MRI Prostate Cancer Reconstruction

### PR DESCRIPTION
**Pull request by Harrison-Cleland**
This pull request implements a VQVAE based architecture for generating synthetic MRI scans of Prostate cancers. The implementation focuses on demonstrating the potential of deep learning medial image generation that is resource efficient and accurate.

* dataset.py: contains functions to pre-process the .nii medial images of the MRI scans
* utils.py: contains methods for data augmentation and optimization
* modules.py: contains the VQVAE architecture and helps classes including an encoding, decoder and optimizer
* predict.py: contains methods for reconstructing images and displaying them
* train.py: the main file of the project that trains the VQVAE model and saves it for later image reconstruction

Please refer to the README file in both md and pdf versions for more information and description of the implementation

Harrison Cleland 47433386

### CODE REVIEW

**Comment by Uto0op** [comment]
Recognition problem: VQVAE on HipMRI ---- Hard Difficulty

1. Algorithm solves the problem:
Clear and resembling reconstructed images and SSIM over 0.7, yet no plots of training/validation loss/SSIM provided -1 (4 Marks)

2. Algorithm demonstrated and functions: 
Lack of validation -0.5 (2.5 Marks)

3. Design:
Modularized design and good programming practice using `utils.py` (1 Mark)

4. Commenting:
Meaningful and detailed comments (1 Mark)

5. Algorithm above Normal Difficulty: (5 Marks)

6. Algorithm is Hard difficulty: (5 Marks)

**Suggestions**:

- README.md could be enriched: Description on the training process should include key hyperparameters and tuning (e.g. why you ended up choosing LR of 1e-3, what else you had attempted to improve the performance and what were their implications); The results should provide reconstructed images (ideally in comparison to the original ones);
- The loss/SSIM during training/validation are expected, otherwise it's hard to see how your model performs/converges;
- Validation set initialized but never used, so is the weight_decay;
- You need to clearly state any reference you used.

**Comment by YashMitttal** [comment]
I would highly suggest you to make your readme more descriptive.

You can do things like:
- Add a Diagram of the Model Architecture
- Provide a brief explanation of what SSIM (Structural Similarity Index) is and why it’s a good measure of image quality
- <<<<<<< HEAD is written at the top of your readme; remove that.
- Include some loss graphs also.

Please focus on the explanation and add some visualisations and references.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. VQVAE (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. Needs more work including loss graphs.</td><td>-2</td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Acceptable. No references.</td><td>-1</td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF submitted.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time submission.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is adequate. </td><td>-1</td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-4</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
File structure not followed, code in parent directory, should be in its own folder. -2

**Comment by shakes76** [comment]
Cant merge because of conflicting changes to main repo files. Please update for merge, doesn't affect grade.
